client/dashboard/me/tax-details/user-tax-form.tsx

```diff
@@ -0,0 +1,410 @@
+			value={ getValue( { item: data } ) || '' }
+		/>
+	);
+}
+
+export default function UserTaxForm() {
+	const { data: countryList } = useSuspenseQuery( countryListQuery() );
+	const countryCodes = getDataFormCountryCodes( countryList );
+
+	const [ localData, setLocalData ] = useState< Partial< UserTaxFormData > >( {} );
+	const query = useSuspenseQuery( userTaxDetailsQuery() );
+	const userTaxDetails: UserTaxDetails = query.data ?? emptyUserTaxDetails;
+	const { data: geoData } = useSuspenseQuery( geoLocationQuery() );
+	const formData = useMemo( () => {
+		const serverData = {
+			country: userTaxDetails.country ?? '',
+			id: userTaxDetails.id ?? '',
+			name: userTaxDetails.name ?? '',
+			address: userTaxDetails.address ?? '',
+		};
+		return {
+			...serverData,
+			...localData,
+		};
+	}, [
+		localData,
+		userTaxDetails?.address,
+		userTaxDetails?.country,
+		userTaxDetails?.id,
+		userTaxDetails?.name,
+	] );
+	const taxName = getTaxName( countryList, formData.country ?? geoData?.country_short ?? 'GB' );
+	const { createSuccessNotice, createErrorNotice, removeNotice } = useDispatch( noticesStore );
+	const userTaxDetailsMutation = () => {
+		return mutationOptions< UserTaxDetails, UserTaxDetailsUpdateError, UserTaxDetails >( {
+			mutationFn: updateUserTaxDetails,
+			onSuccess: ( newData: UserTaxDetails ) => {
+				queryClient.setQueryData(
+					userTaxDetailsQuery().queryKey,
+					( oldData ) => oldData && { ...oldData, ...newData }
+				);
+				removeNotice( 'vat_info_notice' );
+				createSuccessNotice(
+					sprintf(
+						/* translators: %s is the name of taxes in the country (eg: "VAT" or "GST"). */
+						__( 'Your %s details have been updated!' ),
+						taxName ?? __( 'VAT' )
+					),
+					{
+						id: 'vat_info_notice',
+						type: 'snackbar',
+					}
+				);
+			},
+			onError: ( error: Error | UserTaxDetailsUpdateError | UserTaxDetailsFetchError ) => {
+				removeNotice( 'vat_info_notice' );
+				if ( error?.message?.length > 0 ) {
+					createErrorNotice( error.message, { type: 'snackbar', id: 'vat_info_notice' } );
+					return;
+				}
+				createErrorNotice(
+					sprintf(
+						/* translators: %s is the name of taxes in the country (eg: "VAT" or "GST"). */
+						__( 'An error occurred while fetching %s details.' ),
+						taxName ?? __( 'VAT' )
+					),
+					{
+						type: 'snackbar',
+						id: 'vat_info_notice',
+					}
+				);
+			},
+		} );
+	};
+
+	const mutation = useMutation< UserTaxDetails, UserTaxDetailsUpdateError, UserTaxDetails >(
+		userTaxDetailsMutation()
+	);
+	const formatUserTaxDetails = useCallback( ( data: UserTaxDetails ) => {
+		const { country, id } = data;
+
+		if ( !! id && id?.length > 1 ) {
+			return { ...data, id: stripCountryCodeFromVatId( id, country ) };
+		}
+
+		return data;
+	}, [] );
+	const setDetails = useCallback(
+		( userTaxDetails: UserTaxDetails ) => {
+			return mutation.mutate( formatUserTaxDetails( userTaxDetails ) );
```

Comment:
You can actually provide `onSuccess` and `onError` callbacks here too.
That might be the best thing to do in this instance, because we should be defining the `userTaxDetailsMutation` mutation in the `api-queries` package, and that is where it should be updating the query cache based on the data it gets back from the server. But the code for dealing with snackbars etc. should be in this component.

e.g.


```suggestion
			return mutation.mutate( formatUserTaxDetails( userTaxDetails, {
				onSuccess() {
					// Do not put the queryClient.setQueryData() code here
					// It goes in the `api-queries` package
					recordTracksEvent( 'calypso_dashboard_vat_details_validation_success' );
					removeNotice( 'vat_info_notice' );
					// ... etc.
				},
				onError() {
					// ... etc.
				},
			}  ) );
```